x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java

```diff
@@ -141,17 +143,47 @@ private Object unwrapMultiValue(Object values) {
 
     @SuppressWarnings("unchecked")
     Object extractFromSource(Map<String, Object> map) {
-        Object value = map;
-        boolean first = true;
-        // each node is a key inside the map
-        for (String node : path) {
-            if (value == null) {
-                return null;
-            } else if (first || value instanceof Map) {
-                first = false;
-                value = ((Map<String, Object>) value).get(node);
-            } else {
-                throw new SqlIllegalArgumentException("Cannot extract value [{}] from source", fieldName);
+        Object value = null;
+
+        // Used to avoid recursive method calls
+        // Holds the sub-maps in the document hierarchy that are pending to be inspected.
+        LinkedList<Map<String, Object>> queue = new LinkedList<>();
+        // along with the current index of the `path`.
+        LinkedList<Integer> idxQueue = new LinkedList<>();
+
+        queue.add(map);
+        idxQueue.add(-1);
+
+
+        while (!queue.isEmpty()) {
+            int idx = idxQueue.removeLast();
+            Map<String, Object> subMap = queue.removeLast();
+
+            // Find all possible entries by examining all combinations under the current level ("idx") of the "path"
+            // e.g.: If the path == "a.b.c.d" and the idx == 0, we need to check the current subMap against the keys:
+            //       "b", "b.c" and "b.c.d"
+            StringJoiner sj = new StringJoiner(".");
+            for (int i = idx + 1; i < path.length; i++) {
+                sj.add(path[i]);
+                Object node = subMap.get(sj.toString());
+                if (node instanceof Map) {
+                    // Add the sub-map to the queue along with the current path index
+                    queue.add((Map<String, Object>) node);
+                    idxQueue.add(i);
```

Comment:
Why not add i+1 to the queue since the next loop should start from the next position?

Reply:
no reason, I just preferred to start the `i` index of the for loop from prevPosition + 1 just because personally I see it more "visible". Would you prefer it your way?

Reply:
No.